### PR TITLE
Added dependency resolution modes

### PR DESCRIPTION
**Pull request by cjhoward92**
See [issue #11](https://github.com/jeffijoe/awilix/issues/11) for details.

Examples related to transpiling can be found in the examples folder under `typescript` and `babel`.

Code coverage is ~99.62%. I cannot find the last line. `coveralls` was not working for me.

### CODE REVIEW

**Comment by jeffijoe** [comment]
To run the coverage, use `npm run cover`- `npm run coveralls` is used by CI.

**Comment by jeffijoe** [comment]
Here's the culprit: https://coveralls.io/builds/11854857/source?filename=lib%2FcreateContainer.js#L137

You sure its needed?

**Comment by cjhoward92** [comment]
Ah yes. No, that method is no longer needed. I meant to remove it and just forgot. One moment.

**Comment by cjhoward92** [comment]
Updated. You should see 100% coverage come through here soon. Let me know if there is anything you want changed in my implementation.

**Comment by jeffijoe** [comment]
I'm going to review & merge tomorrow, probably edit the readme. I want to make sure we don't overwhelm the first-time reader by explaining all the configuration methods in the introductory example. I think the new Resolution Mode deserves its own section. 😄 

Nice work! 🎊

**Comment by cjhoward92** [comment]
Sounds good to me! Thanks a ton! The code in there is well-written so it was an easy addition.

File: README.md [hunk 1]
```diff
@@ -731,7 +747,7 @@ The `['glob', Lifetime.SCOPED]` syntax is a shorthand for passing in registratio
 
 Clone repo, run `npm i` to install all dependencies, and then `npm run test-watch` + `npm run lint-watch` to start writing code.
 
-For code coverage, run `npm run coverage`.
+For code coverage, run `npm run cover`.
```
Comment (jeffijoe) [inline README.md]: My bad, didn't realize I put `coverage` instead of `cover`.

Comment (cjhoward92) [inline README.md]: It's cool. I figured I'd take care of it for you!

File: lib/loadModules.js [hunk 2]
```diff
@@ -39,7 +40,8 @@ module.exports = function loadModules (dependencies, globPatterns, opts) {
   opts = opts || {}
   opts = Object.assign({}, {
     registrationOptions: {
-      lifetime: Lifetime.TRANSIENT
+      lifetime: Lifetime.TRANSIENT,
+      resolutionMode: ResolutionMode.PROXY
```
Comment (jeffijoe) [inline lib/loadModules.js]: Might want to consider pulling the default from the container.

File: lib/registrations.js [hunk 3]
```diff
@@ -124,15 +134,96 @@ const asClass = (Type, opts) => {
...
+ * Returns a function used to invoke a type constructor.
+ *
+ * @param {Class} Type
+ * The type being contructed.
+ *
+ * @return {Function}
+ * The function that constructs the specified type
+ */
+function buildNewClassFunction (Type) {
+  return (dependencies) => {
```
Comment (jeffijoe) [inline lib/registrations.js]: If anything, you can use `return function () {}` and use `arguments`.

File: lib/registrations.js [hunk 4]
```diff
@@ -124,15 +134,96 @@ const asClass = (Type, opts) => {
...
+ *
+ * @return {Function}
+ * The function used for dependency resolution
+ */
+function generateResolve (fn, depsFn, isFunction) {
+  // Use a regular function instead of an arrow function to facilitate binding to the container object
+  return function (container) {
+    // Because the container holds a global reolutionMode we need to determine it in the proper order or precedence:
+    // registration -> container -> default value
+    let resolutionMode = (this.resolutionMode || container.__resolutionMode || ResolutionModes.PROXY)
```
Comment (jeffijoe) [inline lib/registrations.js]: Maybe we should have a `container.options` where things like the default `resolutionMode` are used, so accessing them don't seem like a hack (the `__` in the name indicates "private"). Thoughts?

**Comment by jeffijoe** [comment]
These comments I just left are things I could take care of, too - not trying to rub all the work off on you! 😄

File: lib/registrations.js (hunk 4 above)
Comment (cjhoward92) [inline lib/registrations.js]: I agree with having `options` on the container. I was thinking along those lines but you weren't storing the options on the container before so I didn't know if that would be too much. The `__` was intentional so hopefully no one would use the property haha.

I will make the other changes as well.

**Comment by cjhoward92** [comment]
@jeffijoe I am in a lucky position at the moment. Tomorrow is my last day at work and they basically said to "Do whatever you want" so I am doing this. I will take care of these issues! It keeps me busy and makes me feel like I am contributing to the community.

**Comment by jeffijoe** [comment]
That sounds great, I'll hold off on the review and edits then! Feel free to hop on the Awilix gitter, I try to be around as much as possible. I'm off for now though, thanks again!

File: lib/registrations.js (hunk 3 above)
Comment (cjhoward92) [inline lib/registrations.js]: I am not sure why I didn't try using `arguments`, it ended up working way better. Thanks!

**Comment by cjhoward92** [comment]
@jeffijoe Updated for review whenever you return.

File: lib/registrations.js [hunk 5]
```diff
@@ -182,27 +166,29 @@ function buildNewClassFunction (Type) {
...
+  if (!dependencyParseTarget) {
+    dependencyParseTarget = fn
+  }
+
   // Use a regular function instead of an arrow function to facilitate binding to the container object
   return function (container) {
     // Because the container holds a global reolutionMode we need to determine it in the proper order or precedence:
     // registration -> container -> default value
-    let resolutionMode = (this.resolutionMode || container.__resolutionMode || ResolutionModes.PROXY)
+    let resolutionMode = (this.resolutionMode || container.options.resolutionMode || ResolutionModes.PROXY)
```
Comment (jeffijoe) [inline lib/registrations.js]: Not sure if this is enforced in the linter, but if you don't modify a variable after initialization, you should use `const`, it communicates intent and the reader will know that it won't be modified. 😄  I can fix these no problem, just letting you know.

**Comment by jeffijoe** [comment]
Made a few modifications + readme updates and other small things, feel free to review my commits! Released as 2.3.0 on npm.

Again, great work @cjhoward92! Are you on Twitter? I wanna give a shoutout. 😄

**Pull request merged**